src/action_exec.ml

```diff
@@ -138,7 +138,7 @@ let rec exec t ~ectx ~dir ~env ~stdout_to ~stderr_to =
       if is_copied_from_source_tree file1 &&
          not (is_copied_from_source_tree file2) then begin
         Promotion.File.register
-          { src = file2
```

Comment:
I'm confused about these action paths. Where are they coming from? How do we know they're build paths here?

Reply:
Well, something would be quite wrong if we tried to promote a file not from the build directory.

I think we'll need to modify the action interface to allow for more than one path type. That will make these _exn calls unnecessary. However, that's quite a larger task.

Reply:
OK, fair enough.
Without that it seems that the benefit of the separate type is not really realized, but one step at a time, I guess.

Reply:
Yes, but I still think this is an improvement. I believe failing here is better than promoting a file from the wrong source. At least the user will catch the bug, rather than just proceeding with the incorrect behavior silently.